File: hw07/hw07/BoardArray2D.java
import java.io.*;
import java.io.IOException;

/**
 * this class is the extends of an AbstractBoard for the game
 * it is has the data 2D array
 */
public class BoardArray2D extends AbstractBoard {
    private int[][] board;
    private int[] space;
    private int size1;
    private int size2;

    /**
     * no parameter constructor
     */
   public BoardArray2D ()
    {
        Number_Of_Boards++;
        space = new int[3];
        int[][] board = new int[9][9];
    }
    public boolean Equals(AbstractBoard otherObject) {
        BoardArray2D newObj = (BoardArray2D) otherObject;
        if(size1 != newObj.size1 || size2 != newObj.size2)
            return false;
        for (int i = 0; i < size1; ++i)
        {
            for (int j = 0; j < size2; ++j)
            {
                if(board[i][j] != newObj.board[i][j])
                    return false;
            }
        }
        if(size1 == newObj.size1 && size2 == newObj.size2)
            return true ;
        else
            return false;
    }

    /**
     * getting the size1
     * @return the size1
     */
    public int getSize1()
    {
        return size1;
    }
    /**
     * getting the size2
     * @return the size2
     */
    public int getSize2()
    {
        return size2;
    }
    /**
     * toString method for the objects
     * @return like a string the object's board
     */
    public String toString()
    {
        String boardStr = "";

        for (int i = 0; i < size1; i++) {
            for (int j = 0; j < size2; j++) {
                if (board[i][j] == -1) {
                   boardStr += "bb ";
                }
                if (board[i][j] < 10 && board[i][j] != -1) {
                    boardStr += String.valueOf(0);
                    boardStr += String.valueOf(board[i][j]);
                    boardStr += " ";
                } else if (board[i][j] != -1) {
                    boardStr += String.valueOf(board[i][j]);
                    boardStr += " ";
                }
            }
            boardStr += "\n";
        }
        return boardStr;
    }

     /**
     * Takes two indexes and returns the corresponding cell content
     * @param i  first index to check
     * @param j second index to check
     * @return the corresponding cell content
     */
    public int cell (int i, int j)
    {

        return (board[i][j]);
    }
    /**
     * Sets the board size to given values. The values are given as parameters
     * and there are no restrictions on the board size
     * @param size1 first value for set
     * @param size2 second value for set
     */
    public void setSize(int size1, int size2)
    {
        board = new int[size1][size2];
        space = new int[2];
    }
    /**
     * Resets the board to the solution.
     */
    public void reset()
    {
        int a=1;
        for (int i = 0; i < size1; ++i)
        {
            for (int j = 0; j < size2; ++j)
            {
                if(board[i][j] != 0)
                {
                    board[i][j] = a;
                    a++;
                }
            }
        }
        board[size1-1][size2-1] = -1;
        space[0]=size1-1;
        space[1]=size2-1;
    }
    /**
     * geting the space index i
     * @return space's i location
     */
    public int getSpacei(){
        return space[0];
    }

    /**
     * geting the space index j
     * @return space's j location
     */
    public int getSpacej(){
        return space[1];
    }
     /**
     * check is the board is a solution
     * @return true if the board is a solution
     */
    public boolean isSolved()
    {
        int a=1;
        //is board final solution
        for(int i=0; i<size1; ++i)
        {
            for(int j=0; j<size2; ++j)
            {
                if((i == size1-1 && j == size2-1 )){}
                else if(board[i][j] == 0){}
                else if(board[i][j]!= a)
                    return false;
                else
                    a++;
            }
        }
        return true;
    }

    /**
     * Reads the board from the file given as function parameter.
     * @param fileName name of file to read
     */
    public void readFromFile(String fileName) {

        int i = 0, j = 0, flag1 = 1;
        int indexBuffer = 0;
        String buffer = new String();

        try {
            FileReader folder = new FileReader(fileName);
            BufferedReader br = new BufferedReader(folder);
            while ((buffer = br.readLine()) != null){
                indexBuffer = 0;
                while (indexBuffer < buffer.length() ) {
                    if (indexBuffer != 0)
                        size2 = j + 1;
                    indexBuffer += 3;
                    j++;
                }
                i++;
                j = 0;
            }

            if (flag1 == 1) {
                size1 = i ;
                flag1 = 0;
            }
            folder.close();

            i = 0;
            j = 0;
            flag1 = 1;
            indexBuffer = 0;
            setSize(size1,size2);

            folder = new FileReader(fileName);
            br = new BufferedReader(folder);

            while ((buffer = br.readLine()) != null) {
                indexBuffer = 0;
                while (indexBuffer < buffer.length()) {
                    if (buffer.charAt(indexBuffer) == 'b') {
                        board[i][j] = -1;
                        space[0] = i;
                        space[1] = j;
                    } else
                        board[i][j] = (buffer.charAt(indexBuffer) - 48) * 10 + buffer.charAt(indexBuffer + 1) - 48;

                    indexBuffer += 3;
                    j++;
                }
                i++;
                j = 0;
            }
            folder.close();
        }
        catch(IOException e) {

            System.out.println("An error occurred.");
            System.exit(1);
        }
    }

    /**
     * writes the board to the file given as function parameter
     * @param fileName name of file to write
     */
    public void writeToFile(String fileName)
    {
        try {
            FileWriter myWriter = new FileWriter(fileName);
            myWriter.write(toString());
            myWriter.close();
        } catch (IOException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }

    /**
     * Makes a move according to the given char parameter
     * @param m is the move route
     * @return 0 if the move is not legal, otherwise 1
     */
    public int move(char m)
    {
        int temp;
        /*swapping the empty cell with direction is choosen using i and j. for all directions*/
        /*if there is no room returns 0*/

        if(m == 'R'){
            if(space[1]+1 == size2 || board[space[0]][space[1]+1] == 0)
                return 0;
            else
            {
                last_Move = 'R';
                number_Of_Moves++;
                temp = board[space[0]][space[1]+1];
                board[space[0]][space[1]+1] = board[space[0]][space[1]];
                board[space[0]][space[1]] = temp;
                space[1]++;
            }
        }
        else if(m == 'L'){
            if(space[1]-1 == -1 || board[space[0]][space[1]-1] == 0)
                return 0;
            else
            {
                last_Move = 'L';
                number_Of_Moves++;
                temp = board[space[0]][space[1]-1];
                board[space[0]][space[1]-1] = board[space[0]][space[1]];
                board[space[0]][space[1]] = temp;
                space[1]--;
            }
        }
        else if(m == 'U'){
            if(space[0]-1 == -1 || board[space[0]-1][space[1]] == 0)
                return 0;
            else{
               last_Move = 'U';
                number_Of_Moves++;
                temp = board[space[0]-1][space[1]];
                board[space[0]-1][space[1]] = board[space[0]][space[1]];
                board[space[0]][space[1]] = temp;
                space[0]--;
            }
        }
        else if(m == 'D'){
            if(space[0]+1 == size1 || board[space[0]+1][space[1]] == 0)
                return 0;
            else
            {
                last_Move = 'D';
                number_Of_Moves++;
                temp = board[space[0]+1][space[1]];
                board[space[0]+1][space[1]] = board[space[0]][space[1]];
                board[space[0]][space[1]] = temp;
                space[0]++;
            }
        }
        return 1;
    }

}

File: hw05/hw05/boardarray2d.cpp
#include "boardarray2d.h"

namespace Game
{
	BoardArray2D :: BoardArray2D ()
	{
		Number_Of_Boards++;
		space = new int[3];

		board = new int*[9];
		for(int i = 0; i < 9; ++i)
			board[i] = new int[9];

	}

	BoardArray2D :: BoardArray2D(const BoardArray2D& boardArrayObject)	//copy constructor
	{

		space = new int[3];
		for(int i = 0; i<3; ++i)
			space[i] = boardArrayObject.space[i];

		size1 = boardArrayObject.size1;
		size2 = boardArrayObject.size2;

		board = new int*[size1];
		for(int i = 0; i < size1; ++i)
			board[i] = new int[size2];

		for (int i = 0; i < size1; ++i)
		{
			for(int j = 0; j < size2; ++j)
			{
				board[i][j] = boardArrayObject.board[i][j];
			}
		}
	}

	BoardArray2D ::  ~BoardArray2D ()
	{
		delete [] space;

		for(int i = 0; i < size1; ++i)
		{
			delete [] board[i];
		}
		delete [] board;
	}

	BoardArray2D& BoardArray2D :: operator = (const BoardArray2D& rightSide)
	{
		size1 = rightSide.size1;
		size2 = rightSide.size2;

		for(int i = 0; i<3; ++i)
			space[i] = rightSide.space[i];
		
		for (int i = 0; i < size1; ++i)
		{
			for(int j = 0; j < size2; ++j)
			{
				board[i][j] = rightSide.board[i][j];
			}
		}

		return *this;
	}

	void BoardArray2D::print() const{
		for(int i=0; i<size1; i++)
		{
			for(int j=0; j<size2; j++)
			{
				if(board[i][j] == -1)
					cout<<"bb	";
				if(board[i][j] < 10 && board[i][j] != -1)
					cout<< 0 << board[i][j] << "	";
				else if(board[i][j] != -1)
					cout<< board[i][j] << "	";
			}
			cout<< endl;
		}
	}

	void BoardArray2D::readFromFile(const string fileName)
	{
		ifstream folder;

		auto i=0, j=0,flag1=1;
		auto indexBuffer=0;									
		string buffer;
		
		folder.open(fileName.c_str());
		do
		{
			getline (folder, buffer);
			indexBuffer = 0;
			while(indexBuffer<buffer.size()){			
				if(indexBuffer != 0)
					size2 = j+1;
				indexBuffer += 3;
				j++;
			}
			i++;
			j=0;
		}while(!folder.eof());

		if(flag1){
			size1 = i-1;
			flag1 =0; 
		}
		folder.close();
	
		i=0, j=0,flag1=1;
		indexBuffer=0;
		
		setSize(size1,size2);
		
		folder.open(fileName.c_str());
		do
		{
			getline (folder, buffer);
			indexBuffer = 0;
			while(indexBuffer<buffer.size()){
				if(buffer.at(indexBuffer) == 'b'){
					board[i][j] = -1;
					space[0] = i;
					space[1] = j;
				}
				else
					board[i][j] = (buffer.at(indexBuffer)-48) * 10 + buffer.at(indexBuffer+1)-48;

				indexBuffer += 3;
				j++;
			}
			i++;
			j=0;
		}while(!folder.eof());

		folder.close();
	}

	void BoardArray2D::writeToFile(const string fileName) const
	{
		ofstream folder1;
		folder1.open(fileName.c_str());
		for(int i=0; i<size1; i++)
		{
			for(int j=0; j<size2; j++)
			{	
				if(i == size1-1 && j == size2-1){
					if(board[i][j] == -1)
						folder1<<"bb";
					else if(board[i][j] < 10)
						folder1<< 0<< board[i][j];
					else
						folder1<<board[i][j];
				}

				else
				{
					if(board[i][j] == -1)
						folder1<<"bb	";
					else if(board[i][j] < 10)
						folder1<< 0 << board[i][j] << "	";
					else if(board[i][j] != -1)
						folder1<< board[i][j] << "	";
				}
			}
			folder1 << endl;
		}
	}

	void BoardArray2D::reset()
	{
		auto a=1;	
		for (int i = 0; i < size1; ++i)
		{
			for (int j = 0; j < size2; ++j)
			{
				if(board[i][j] != 0)
				{
					board[i][j] = a;
					a++;
				}
				
			}
			cout << endl;
		}
		board[size1-1][size2-1] = -1;
		print();
		space[0]=size1-1;
		space[1]=size2-1;
	}
	void BoardArray2D::setSize(int size1, int size2)
	{
		for(int i = 0; i < 9; ++i)
		{
			delete [] board[i];
		}
		delete [] board;
		board = new int*[size1];
		for(int i = 0; i < size1; ++i)
			board[i] = new int[size2];

	}

	int BoardArray2D::move(char m)
	{
		int temp;
		/*swapping the empty cell with direction is choosen using i and j. for all directions*/
		/*if there is no room returns 0*/

		if(m == 'R'){	
			if(space[1]+1 == size2 || board[space[0]][space[1]+1] == 0)	
				return 0;
			else	
			{
				last_Move = 'R';
				number_Of_Moves++;
				temp = board[space[0]][space[1]+1];
				board[space[0]][space[1]+1] = board[space[0]][space[1]];
				board[space[0]][space[1]] = temp;
				space[1]++;	
			}		
		}
		else if(m == 'L'){
			if(space[1]-1 == -1 || board[space[0]][space[1]-1] == 0)
				return 0;
			else
			{
				last_Move = 'L';
				number_Of_Moves++;
				temp = board[space[0]][space[1]-1];
				board[space[0]][space[1]-1] = board[space[0]][space[1]];
				board[space[0]][space[1]] = temp;
				space[1]--;
			}
		}
		else if(m == 'U'){
			if(space[0]-1 == -1 || board[space[0]-1][space[1]] == 0)
				return 0;
			else{
				last_Move = 'U';
				number_Of_Moves++;
				temp = board[space[0]-1][space[1]];
				board[space[0]-1][space[1]] = board[space[0]][space[1]];
				board[space[0]][space[1]] = temp;
				space[0]--;
			}
		}
		else if(m == 'D'){
			if(space[0]+1 == size1 || board[space[0]+1][space[1]] == 0)
				return 0;
			else
			{
				last_Move = 'D';
				number_Of_Moves++;
				temp = board[space[0]+1][space[1]];
				board[space[0]+1][space[1]] = board[space[0]][space[1]];
				board[space[0]][space[1]] = temp;
				space[0]++;
			}
		}
		return 1;
	}

	bool BoardArray2D::isSolved() const
	{
		int a=1;
		//is board final solution
		for(int i=0; i<size1; ++i)
		{
			for(int j=0; j<size2; ++j)
			{
				if((i == size1-1 && j == size2-1 )){}
				else if(board[i][j] == 0){}
				else if(board[i][j]!= a)
					return false;
				else
					a++;		
			}
		}
		return true;
	}

	int BoardArray2D::operator () (int i, int j) const
	{
		return (board[i][j]);
	}

	bool BoardArray2D::operator == (const BoardArray2D& other) 
	{
		for (int i = 0; i < size1; ++i)
		{
			for (int j = 0; j < size2; ++j)
			{
				if(operator()(i,j) != other.operator()(i,j))
						return false;
			}
		}
		if(size1 == other.size1 && size2 == other.size2)
			return true ;
		else
			return false;
	}

}